alibi/explainers/similarity/backends/pytorch/base.py

```diff
@@ -0,0 +1,97 @@
+"""Torch backend for similarity explainers.
+
+Methods unique to the Torch backend are defined here. The interface this class defines syncs with the torch backend in
+order to ensure that the similarity methods only require to match this interface.
+"""
+
+import random
+from typing import Callable
+
+import numpy as np
+import torch.nn as nn
+import torch
+
+
+class _TorchBackend(object):
+    device = None
+
+    @staticmethod
+    def get_grads(
+            model: nn.Module,
+            x: torch.Tensor,
+            y: torch.Tensor,
+            loss_fn: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    ) -> np.ndarray:
+        """
+        Computes the gradients of the loss function with respect to the model's parameters for a single training and
+        target pair.
+
+        Parameters:
+        -----------
+        model:
+            The model to compute gradients for.
+        x:
+            The input data.
+        y:
+            The target data.
+        loss_fn:
+            The loss function to use.
+
+        Returns:
+        --------
+        grads:
+            The gradients of the loss function with respect to the model's parameters. This is returned as a flattened \
+            array.
+        """
+
+        for param in model.parameters():
+            if isinstance(param.grad, torch.Tensor):
+                param.grad.data.zero_()
+
+        output = model(x)
+        loss = loss_fn(output, y)
+        loss.backward()
```

Comment:
Make sure that the model is in `eval` state. Some layers as `Dropout`, `BatchNorm` etc have different behaviour when in `train` vs `eval`.
You need to add `model.eval()`.
Does it make sense to support gradients for `eval` and `train` by passing an argument?

Reply:
Yeah, good point. I think it probably makes sense to revert the model to training mode once the gradients have been computed. That way the user gets the same model back as they put in?

Reply:
Is there a use case where you'd want to compute the gradients with the dropout?

Reply:
Actually having thought about it this is a little tricky because whatever we do we need to revert the state and make it clear what's happening. If the user specifies they want to use the explainer in eval mode but pass it in training mode then we should change it to eval and then change it back at the end otherwise we're setting model state within the explainer which seems weird. What if they pass it in eval but want to evaluate it in training? It seems weird that we'd have a parameter to manage that option? 

Ideally, the user would set themselves and we make clear they need to do so in the docs but this isn't possible because it doesn't work with the Keras API which expects `training` to be passed as a parameter to the model `__call__` method.

Options:
2. If there is just no use case for doing the above in training mode we can just set eval and then revert after.
3. Otherwise we pass a parameter that's either training or eval, we then change the model to that state during and then revert it to whatever state they initially passed it in.

Reply:
I think 1. is fine unless @RobertSamoilescu disagrees.

Reply:
I also think that 1 is fine for now. Cannot come up with any example in which we need the model in training mode.